packages/application-components/src/components/custom-views/custom-view-loader/custom-view-loader.tsx

```diff
@@ -0,0 +1,142 @@
+import messages from './messages';
+
+type TCustomViewIframeMessage = {
+  source: string;
+  destination: string;
+  eventName: string;
+  eventData: Record<string, unknown>;
+};
+
+/*
+  TODO: These types are temporary until we have the proper
+  ones generated from the Settings schema
+*/
+export type TPermissionGroup = {
+  name: string;
+  oAuthScopes: string[];
+};
+
+export type TCustomView = {
+  id: string;
+  defaultLabel: string;
+  labelAllLocales: Record<string, string>;
+  url: string;
+  type: 'custom-panel' | 'custom-tab';
+  typeConfig?: {
+    size?: 'small' | 'large';
+  };
+  locators: string[];
+};
+
+type TCustomViewLoaderProps = {
+  customView: TCustomView;
+  onClose: () => void;
+};
+
+const isIframeReady = (iFrameElementRef: HTMLIFrameElement) => {
+  try {
+    return iFrameElementRef?.contentWindow?.document.readyState === 'complete';
+  } catch {
+    // Trying to access the contentWindow of a cross-origin iFrame will throw an error.
+    // We are not supposed to even get here because the iFrame must use
+    // a URL from our very same domain (the custom view is proxied through our http-proxy service).
+    return false;
+  }
+};
+
+const CustomPanelIframe = styled.iframe`
+  height: 100%;
+  width: 100%;
+  border: none;
+`;
+
+function CustomViewLoader(props: TCustomViewLoaderProps) {
+  const iFrameElementRef = useRef<HTMLIFrameElement>(null);
+  const appContext = useApplicationContext();
+  const intercomChannel = useRef(new MessageChannel());
+  const showNotification = useShowNotification();
+  const intl = useIntl();
+
+  const messageFromIFrameHandler = useCallback((event: MessageEvent) => {
+    if (event.data.origin === window.location.origin) {
+      console.log('message received from iframe port: ', event);
+    }
+  }, []);
+
+  // onLoad handler is called from the iFrame even where the URL is not valid
+  // (blocked by CORS, 404, etc.) so we need to make sure the iFrame is ready
+  const onLoadSuccessHandler = useCallback(() => {
+    // Show error and block if the iFrame is not ready
+    // (error loading it)
+    if (!isIframeReady(iFrameElementRef.current!)) {
+      showNotification({
+        domain: DOMAINS.PAGE,
+        kind: NOTIFICATION_KINDS_PAGE.error,
+        text: intl.formatMessage(messages.loadError),
+      });
+      return;
+    }
+
+    // Listen for messages from the iFrame
+    intercomChannel.current.port1.onmessage = messageFromIFrameHandler;
+
+    // Transfer port2 to the iFrame so it can send messages back privately
+    iFrameElementRef.current?.contentWindow?.postMessage(
+      'mc_init',
+      window.location.href,
+      [intercomChannel.current.port2]
+    );
+
+    // Send the initialization message to the iFrame
+    intercomChannel.current.port1.postMessage({
+      source: 'mc-host-application',
+      destination: `custom-view-${props.customView.id}`,
+      eventName: 'hostAcknowledgesIframeLoaded',
+      eventData: {
+        appContext,
```

Comment:
Currently we will only need the projectKey and the data language in the iFrame, but I thought passing all the application context just in case.
Let me know if you think we should be more specific.

Reply:
Ok. 

Updated here: 041828a83cab9f34b7a92a11a099dcd6546c3369